pkgs/development/python-modules/gpuctypes/default.nix

```diff
@@ -0,0 +1,91 @@
+{ lib
+, config
+, buildPythonPackage
+, fetchFromGitHub
+, substituteAll
+, addDriverRunpath
+, cudaSupport ? config.cudaSupport
+, rocmSupport ? config.rocmSupport
+, cudaPackages
+, setuptools
+, ocl-icd
+, rocmPackages
+, pytestCheckHook
+}:
+buildPythonPackage rec {
+  pname = "gpuctypes";
+  version = "0.3.0";
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    repo = "gpuctypes";
+    owner = "tinygrad";
+    rev = "refs/tags/${version}";
+    hash = "sha256-xUMvMBK1UhZaMZfik0Ia6+siyZGpCkBV+LTnQvzt/rw=";
+  };
+
+  patches = [
+    (substituteAll {
+      src = ./0001-fix-dlopen-cuda.patch;
+      inherit (addDriverRunpath) driverLink;
+      libnvrtc =
+        if cudaSupport
+        then "${lib.getLib cudaPackages.cuda_nvrtc}/lib/libnvrtc.so"
+        else "Please import nixpkgs with `config.cudaSupport = true`";
+    })
+  ];
+
+  nativeBuildInputs = [
+    setuptools
+  ];
+
+  postPatch = ''
+    substituteInPlace gpuctypes/opencl.py \
+      --replace "ctypes.util.find_library('OpenCL')" "'${ocl-icd}/lib/libOpenCL.so'"
+  ''
+  # hipGetDevicePropertiesR0600 is a symbol from rocm-6. We are currently at rocm-5.
+  # We are not sure that this works. Remove when rocm gets updated to version 6.
+  + lib.optionalString rocmSupport ''
+    substituteInPlace gpuctypes/hip.py \
+      --replace "/opt/rocm/lib/libamdhip64.so" "${rocmPackages.clr}/lib/libamdhip64.so" \
+      --replace "/opt/rocm/lib/libhiprtc.so" "${rocmPackages.clr}/lib/libhiprtc.so" \
+      --replace "hipGetDevicePropertiesR0600" "hipGetDeviceProperties"
```

Comment:
One more thought: this is likely going to break with `rocmPackages_6`, and this also means this patch doesn't account for the possibility of `.override { rocmPackages = somethingElse; }`. The failure isn't going to be silent (we know the tests fail unless the symbol is found) so that's enough to get started

Reply:
@NixOS/rocm-maintainers could you actually please test this? We've no idea if this works. We could mark this as `broken = rocmSupport` but then we'll most likely never find out if it works because nobody would care to override and test it...

Reply:
I've ran into issues with just patching it like this, the patch is slightly more complicated since the `hipDeviceProp_t` struct changes between rocm 5 and rocm 6, which is why they have the R0600 suffix. I've patched it by just adding the autogenned struct from rocm 5 and replacing `hipDeviceProp_tR0600` with `hipDeviceProp_t`

Reply:
@wozeparrot so does something need to be changed? By the way, if you've got access to amd hardware, could you maybe run the tests?

Reply:
Currently I'm running this patch: https://github.com/wozeparrot/tinygrad-nix/blob/main/hip.patch, but this is against tinygrad master where gpuctypes is no longer used but the patch should apply to gpuctypes as well.

> By the way, if you've got access to amd hardware, could you maybe run the tests?

Yea I can give running the tests a try.